// ts-mockito/lib/MethodAction.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MethodAction = (function () {
    function MethodAction(methodName, args) {
        this.methodName = methodName;
        this.args = args;
        this.callIndex = ++MethodAction.globalCallIndex;
    }
    MethodAction.prototype.isApplicable = function (methodName, matchers) {
        if (this.methodName != methodName) {
            return false;
        }
        var allValid = true;
        var index = 0;
        for (var _i = 0, _a = this.args; _i < _a.length; _i++) {
            var arg = _a[_i];
            if (matchers[index] && !matchers[index].match(arg)) {
                allValid = false;
            }
            index++;
        }
        return allValid;
    };
    MethodAction.prototype.getCallIndex = function () {
        return this.callIndex;
    };
    MethodAction.prototype.hasBeenCalledBefore = function (action) {
        return this.getCallIndex() < action.getCallIndex();
    };
    return MethodAction;
}());
MethodAction.globalCallIndex = 0;
exports.MethodAction = MethodAction;
